src/mapstorage/XmlMapStorage.cpp

```diff
@@ -0,0 +1,462 @@
+        filter.prepare(progressCounter);
+    }
+    auto saveOne = [&stream](const Room &room) { saveRoom(stream, room); };
+    for (const auto &pRoom : roomList) {
+        filter.visitRoom(deref(pRoom), baseMapOnly, saveOne);
+        progressCounter.step();
+    }
+}
+
+void XmlMapStorage::saveRoom(QXmlStreamWriter &stream, const Room &room)
+{
+    stream.writeStartElement("room");
+
+    stream.writeAttribute("id", QString("%1").arg(room.getId().asUint32()));
+    stream.writeAttribute("name", room.getName().toQString());
+    if (!room.isUpToDate()) {
+        stream.writeAttribute("uptodate", "false");
+    }
+    saveXmlElement(stream, "align", alignName(room.getAlignType()));
+    saveXmlElement(stream, "light", lightName(room.getLightType()));
+    saveXmlElement(stream, "portable", portableName(room.getPortableType()));
+    saveXmlElement(stream, "ridable", ridableName(room.getRidableType()));
+    saveXmlElement(stream, "sundeath", sundeathName(room.getSundeathType()));
+    saveXmlElement(stream, "terrain", terrainName(room.getTerrainType()));
+    saveCoordinate(stream, "coord", room.getPosition());
+    saveRoomLoadFlags(stream, room.getLoadFlags());
+    saveRoomMobFlags(stream, room.getMobFlags());
+
+    for (const ExitDirEnum dir : ALL_EXITS7) {
+        saveExit(stream, room.exit(dir), dir);
+    }
+    saveXmlElement(stream, "description", room.getDescription().toQString());
+    saveXmlElement(stream, "contents", room.getContents().toQString());
+    saveXmlElement(stream, "note", room.getNote().toQString());
+
+    stream.writeEndElement(); // end room
+}
+
+void XmlMapStorage::saveCoordinate(QXmlStreamWriter &stream,
+                                   const QString &name,
+                                   const Coordinate &pos)
+{
+    stream.writeStartElement(name);
+    stream.writeAttribute("x", QString("%1").arg(pos.x));
+    stream.writeAttribute("y", QString("%1").arg(pos.y));
+    stream.writeAttribute("z", QString("%1").arg(pos.z));
+    stream.writeEndElement(); // end coord
+}
+
+void XmlMapStorage::saveExit(QXmlStreamWriter &stream, const Exit &e, ExitDirEnum dir)
+{
+    if (!e.exitIsExit() || e.outIsEmpty()) {
+        return;
+    }
+    stream.writeStartElement("exit");
+    stream.writeAttribute("dir", lowercaseDirection(dir));
+    saveExitTo(stream, e);
+    saveXmlAttribute(stream, "doorname", e.getDoorName().toQString());
+    saveDoorFlags(stream, e.getDoorFlags());
+    saveExitFlags(stream, e.getExitFlags());
+    stream.writeEndElement(); // end exit
+}
+
+void XmlMapStorage::saveExitTo(QXmlStreamWriter &stream, const Exit &e)
+{
+    QString idlist;
+    const char *separator = nullptr;
+    for (const RoomId id : e.outRange()) {
+        if (separator == nullptr) {
+            separator = ",";
+        } else {
+            idlist += separator;
+        }
+        idlist += QString("%1").arg(id.asUint32());
+    }
+    stream.writeAttribute("to", idlist);
```

Comment:
Just make "to" an element?

Reply:
And repeat &lt;to&gt;roomId&lt;/to&gt; for each connected room?
That's surely possible, but again quite redundant.
Also, one exit connecting to multiple rooms is rare, basically only happens in random mazes.

It could be useful if the DTD or XML Schema allows XML validator software to cross-reference exits with room ids, i.e. to automatically check if the target room id really exists in the map.
But the only DTD type that allows such a feature is the ID type, which cannot be numeric.

All in all, I'm not sure the extra verbosity is worth it in this case... but if you want I'll make the change

Reply:
Yeah, those kind of things probably don't belong in a DTD anyway. We can do that during map load time.

Reply:
Same as above, "to" is now a repeatable element.